code/game/machinery/clonepod.dm

```diff
@@ -405,127 +582,81 @@ GLOBAL_LIST_INIT(cloner_biomass_items, list(\
+		ui = new(user, src, "CloningPod", "Cloning Pod")
+		ui.open()
 
-	connected.temp = "[name] : [message]"
-	connected.updateUsrDialog()
-	return TRUE
+/obj/machinery/clonepod/ui_data(mob/user)
+	var/list/data = list()
+	data["biomass"] = biomass
+	data["biomass_storage_capacity"] = biomass_storage_capacity
+	data["sanguine_reagent"] = reagents.get_reagent_amount("sanguine_reagent")
+	data["osseous_reagent"] = reagents.get_reagent_amount("osseous_reagent")
 
-/obj/machinery/clonepod/proc/go_out()
-	countdown.stop()
-	var/turf/T = get_turf(src)
-	if(mess) //Clean that mess and dump those gibs!
-		for(var/i in missing_organs)
-			var/obj/I = i
-			I.forceMove(T)
-		missing_organs.Cut()
-		mess = FALSE
-		new /obj/effect/gibspawner/generic(get_turf(src), occupant)
-		playsound(loc, 'sound/effects/splat.ogg', 50, 1)
-		update_icon()
-		return
+	var/list/organs_list
+	for(var/obj/item/organ/O in contents)
+		organs_list += list(list("name" = O.name, "ref" = O.UID()))
 
-	if(!occupant)
-		return
+	data["organs"] = organs_list
+	data["currently_cloning"] = currently_cloning
 
-	if(grab_ghost_when == CLONER_MATURE_CLONE)
-		UnregisterSignal(clonemind.current, COMSIG_LIVING_REVIVE)
-		UnregisterSignal(clonemind, COMSIG_MIND_TRANSER_TO)
-		clonemind.transfer_to(occupant)
-		occupant.grab_ghost()
-		to_chat(occupant, "<span class='userdanger'>You remember nothing from the time that you were dead!</span>")
-		to_chat(occupant, "<span class='notice'><b>There is a bright flash!</b><br>\
-			<i>You feel like a new being.</i></span>")
-		if(HAS_TRAIT(src, TRAIT_CMAGGED))
-			playsound(loc, 'sound/items/bikehorn.ogg', 50, 1)
-			occupant.dna.SetSEState(GLOB.clumsyblock, TRUE, FALSE)
-			occupant.dna.SetSEState(GLOB.comicblock, TRUE, FALSE)
-			singlemutcheck(occupant, GLOB.clumsyblock, MUTCHK_FORCED)
-			singlemutcheck(occupant, GLOB.comicblock, MUTCHK_FORCED)
-			occupant.dna.default_blocks.Add(GLOB.clumsyblock) //Until Genetics fixes you, this is your life now
-			occupant.dna.default_blocks.Add(GLOB.comicblock)
-		occupant.flash_eyes(visual = 1)
-		clonemind = null
-
-
-	QDEL_LIST_CONTENTS(missing_organs)
-	occupant.SetLoseBreath(0) // Stop friggin' dying, gosh damn
-	occupant.setOxyLoss(0)
-	for(var/datum/disease/critical/crit in occupant.viruses)
-		crit.cure()
-	occupant.forceMove(T)
-	occupant.update_body()
-	domutcheck(occupant) //Waiting until they're out before possible notransform.
-	occupant.special_post_clone_handling()
-	occupant = null
-	update_icon()
+	return data
+
+/obj/machinery/clonepod/ui_act(action, params, datum/tgui/ui)
+	if(..())
+		return
+	switch(action)
+		if("eject_organ")
+			var/obj/item/organ/O = locateUID(params["organ_ref"])
+			if(!O) //This shouldn't happen
+				return FALSE
+			if(!ui.user.put_in_hands(O))
+				O.forceMove(loc)
+			return TRUE
+		if("purge_reagent")
+			reagents.del_reagent(params["reagent"])
+			return TRUE
+		if("remove_reagent")
+			reagents.remove_reagent(params["reagent"], text2num(params["amount"]))
```

Comment:
check it's in the list, you also won't need the `text2num` as TGUIv4 will pass numbers as ints natively